### PR TITLE
fix(vercel): handle dots in source path for redirects

### LINKED ISSUE #9260: Vercel adapter redirects generate incorrect routes
### Astro Info

```block
Astro                    v3.6.0
Node                     v18.19.0
System                   macOS (arm64)
Package Manager          pnpm
Output                   static
Adapter                  @astrojs/vercel
Integrations             @astrojs/tailwind
                         @astrojs/react
                         @astrojs/sitemap
                         @astrojs/mdx
```

### If this issue only occurs in one browser, which browser is a problem?

Chrome

### Describe the Bug

I've [configuring redirects](https://docs.astro.build/en/core-concepts/routing/#configured-redirects), the `@astrojs/vercel` will generate "routes" in vercel-build-api file `.vercel/output/config.json`, but it seems that each `route.src` is incorrect. 

The redirects worked on astro dev server, but it's not working on vercel.app (return 404).

My `astro.config.ts`

```typescript
import mdx from "@astrojs/mdx";
import react from "@astrojs/react";
import sitemap from "@astrojs/sitemap";
import tailwind from "@astrojs/tailwind";
import { defineConfig } from "astro/config";
import rehypeSlug from "rehype-slug";
import { SITE } from "./src/config";

import vercel from "@astrojs/vercel/static";

// https://astro.build/config
export default defineConfig({
  site: SITE.website,
  integrations: [
    tailwind({
      applyBaseStyles: false,
    }),
    react(),
    sitemap(),
    mdx(),
  ],
  markdown: {
    shikiConfig: {
      theme: "one-dark-pro",
      wrap: true,
    },
  },
  vite: {
    optimizeDeps: {
      exclude: ["@resvg/resvg-js"],
    },
  },
  scopedStyleStrategy: "where",
  output: "static",
  adapter: vercel({
    webAnalytics: {
      enabled: true,
    },
    speedInsights: {
      enabled: true,
    },
  }),
  redirects: {
    "Basic/http-2-0.html": "/posts/http2",
  },
});
```

The `.vercel/output/config.json`

```json
{
	"version": 3,
	"routes": [
		{
			"src": "/\\/Basic\\/http-2-0\\.html",
			"headers": {
				"Location": "/posts/http2"
			},
			"status": 301
		},
		{
			"src": "^/_astro/(.*)$",
			"headers": {
				"cache-control": "public, max-age=31536000, immutable"
			},
			"continue": true
		},
		{
			"handle": "filesystem"
		}
	]
}
```

### What's the expected result?

redirect works on vercel

### Link to Minimal Reproducible Example

N/A

### Participation

- [ ] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by lilnasy**
## Changes

- Closes #9260

## Testing
Added a test case

## Docs
Does not affect behavior

### CODE REVIEW

File: packages/integrations/vercel/src/lib/redirects.ts [hunk 1]
```diff
@@ -66,7 +65,12 @@ function getRedirectLocation(route: RouteData, config: AstroConfig): string {
 	} else if (typeof route.redirect === 'object') {
 		return pathJoin(config.base, route.redirect.destination);
 	} else {
-		return pathJoin(config.base, route.redirect || '');
+		try {
+			const url = new URL(route.redirect!);
+			return url.href;
+		} catch {
```
Comment (ematipico) [inline packages/integrations/vercel/src/lib/redirects.ts:71]: What's the `try` for? The absence of `.redirect` or a parsing error or `URL`?

File: packages/integrations/vercel/src/lib/redirects.ts [hunk 2]
```diff
@@ -21,8 +21,7 @@ function getMatchPattern(segments: RoutePart[][]) {
 		.map((segment) => {
 			return segment[0].spread
 				? '(?:\\/(.*?))?'
-				: '\\/' +
-						segment
+				: segment
```
Comment (ematipico) [inline packages/integrations/vercel/src/lib/redirects.ts:24]: What's this change for?

File: packages/integrations/vercel/test/redirects.test.js [hunk 3]
```diff
@@ -32,23 +34,41 @@ describe('Redirects', () => {
...
 	});
 
+	it('define external routes', async () => {
+		const config = await getConfig();
+
+		const externalRoute = config.routes.find((r) => r.src === '/external');
+		expect(externalRoute).to.not.be.undefined;
+		expect(externalRoute.headers.Location).to.equal('https://example.com/');
+		expect(externalRoute.status).to.equal(301);
+	});
```
Comment (ematipico) [inline packages/integrations/vercel/test/redirects.test.js:57]: As discussed in `adapters` repo, this shouldn't be a valid case. We don't support external redirects... right?

File: packages/integrations/vercel/src/lib/redirects.ts (hunk 1 above)
Comment (lilnasy) [reply packages/integrations/vercel/src/lib/redirects.ts:71]: If the argument is not a full URL, the constructor throws, which here means we fallback to the regular processing of the path as internal.

This is outdated, as this PR is based on top of the one you reviewed earlier: #9287. It should not process external links.

**Review by matthewp — APPROVED** [review]

**Review by natemoo-re — APPROVED** [review]

**Pull request merged**